ruby/lib/google/protobuf/repeated_field.rb

```diff
@@ -29,11 +29,141 @@
+require 'forwardable'
 module Google
   module Protobuf
     class RepeatedField
+      extend Forwardable
 
-      alias_method :size, :length
+      # methods defined in C or Java:
+      #   +
+      #   [], at
+      #   []=
+      #   clear
+      #   dup, clone
+      #   each
+      #   each_with_index
+      #   hash
+      #   insert
+      #   push, <<
+      #   pop
+      #   replace
+      #   length, size
+      #   ==
+      #   to_ary, to_a
+      #   also all enumerable
+
+      # NOTE:  using delegators rather than method_missing to make the
+      #        relationship explicit instead of implicit
+      # NOTE2: use to_a instea
+      def_delegators :to_ary,
+        :&, :*, :-, :'<=>',
+        :assoc, :bsearch, :combination, :compact, :count, :cycle,
+        :drop, :drop_while, :eql?, :fetch, :find_index, :flatten,
+        :include?, :index, :inspect, :join,
+        :pack, :permutation, :product, :pretty_print, :pretty_print_cycle,
+        :rassoc, :repeated_combination, :repeated_permutation, :reverse, :rindex, :rotate,
+        :sample, :shelljoin, :slice, :to_s, :transpose, :uniq, :|
+
+      def_delegators :to_a, :shuffle
+
+
+      def first(n=nil)
+        n ? self[0..n] : self[0]
+      end
+
+
+      def last(n=nil)
+        n ? self[(self.size-n-1)..-1] : self[-1]
+      end
+
+
+      def pop(n=nil)
+        if n
+          results = []
+          n.times{ results << pop_internal }
+          return results
+        else
+          return pop_internal
+        end
+      end
+
+
+      def empty?
+        self.size == 0
+      end
+
+      # array aliases into enumerable
+      alias_method :each_index, :each_with_index
+      alias_method :slice, :[]
+      alias_method :values_at, :select
+      alias_method :map, :collect
+
+
+      %w(delete_at delete_if shift slice! unshift).each do |method_name|
+        define_method(method_name) do |*args|
+          arr = self.to_a
+          result = arr.send(method_name, *args)
+          self.replace(arr)
+          return result
+        end
+      end
+
+
+      %w(delete).each do |method_name|
+        define_method(method_name) do |*args, &block|
+          arr = self.to_a
+          result = arr.send(method_name, *args)
+          self.replace(arr)
+          return result if result
+          return block ? block.call : result
+        end
+      end
+
+
+      %w(collect! compact! fill flatten! insert reverse! rotate! select! shuffle! sort! sort_by! uniq!).each do |method_name|
+        define_method(method_name) do |*args, &block|
+          # result can be an Enumerator, Array, or nil
+          # Enumerator can sometimes be returned if a block is an optional argument and it is not passed in
+          # nil usually specifies that no change was made
+          result = self.to_a.send(method_name, *args, &block)
+          if result
+            new_arr = result.to_a
+            self.replace(new_arr)
+            if result.is_a?(Enumerator)
```

Comment:
that is what it is doing; it absorbs the changes first, then passes back an enum.  BUT, that enum can make changes, and we need to have a way to also absorb those changes.  See  `arr.collect!.with_index{|x, i| x[0...i] }`.  That is why there is the proxyEnum.  If we didn't have to do that, then `result.to_enum` would be perfect